docs/specs/00001-expectation-system.md

```diff
@@ -0,0 +1,150 @@
+```
+
+And a successful output to that command may look like:
+
+```json
+{
+  "fqdns": "",
+  "id": "/subscriptions/<subscription-id>/resourceGroups/<resource-group>/providers/Microsoft.Compute/virtualMachines/<vm-name>",
+  "identity": {
+    "systemAssignedIdentity": "<managed-identity>",
+    "userAssignedIdentities": {}
+  },
+  "location": "eastus",
+  "macAddress": "00-0D-3A-1C-6B-66",
+  "powerState": "VM running",
+  "privateIpAddress": "10.0.0.4",
+  "publicIpAddress": "172.178.12.226",
+  "resourceGroup": "<resource-group>",
+  "zones": ""
+}
+```
+
+Looking throughout the output, we can see that there is a significant amount of
+variation that can occur between mutliple successful runs of the previous
+command. However though, most of the time we would only like to specify that the
+output either looks like an output that we've seen before or that a
+certain pattern can be found the output itself. In order for the
+Innovation Engine to be able to solve this problem, we need to create a system
+that allows authors to quantify what they expect the result of a command to
+look like.
+
+## Solution
+
+Most of the codeblocks that the Innovation Engine currently executes are written
+in bash and run executables to carry out tasks on behalf of a user. Instead of
+only relying on the exit code to determine that a codeblock successfully
+executed, we must look at the side effects (output to stdout/stderr) produced
+by the command too. We will only perform comparisons when a command succeeds
+and only from output sent to the standard output stream. If documentation
+authors would like to make comparisons using the comamnd output from standard
+error, they can manually combine the standard output/error stream into the
+output stream within the codeblock itself.
+
+To allow documentation authors to set their expectations around results,
+we would like to introduce a new syntax to markdown documents in the form of
+`expectation` tags. Expectation tags are HTML 5 comments that must be defined
+inside of the markdown documents above a result block and below a codeblock.
+Here's how that would look:
+
+````markdown
+# Example codeblock
+
+```bash
+echo "foo"
+```
+
+<!--expectation type="similarity" value="0.7"-->
+
+```
+foo
+```
+````
+
+IE takes the standard output from the execution of the 
+codeblock containing `echo "foo"` and compares it against the result codeblock
+below our expectation tag using [Jaro Similarity](https://en.wikipedia.org/wiki/Jaro%E2%80%93Winkler_distance)
+as the metric for determining the two strings similarity. The value provided by
+the author sets the threshold for how similar the strings must be on a scale
+of 0 to 1.
+
+If a documentation author instead wanted to ensure that a specific pattern
+existed within result of a command execution, they can change the parameters
+`type` and `value` like so:
+
+````markdown
+# Example codeblock
+
+```bash
+echo "foo"
+
+```
+
+<!--expectation type="matches" value="foo" -->
```

Comment:
How useful is an exact match? Is that the intent here? I would have though "contains" would be more useful.

Reply:
`matches` isn't for exact matching, but it is used for matching patterns in the codeblock (Which just so happens to be an exact match here). However, I can see the confusion that calling it matches may create so I will revert back to just naming it `contains`